### PR TITLE
Implement tag interactions

### PR DESCRIPTION
**Pull request by quadrismegistus**
This PR implements what @rlskoeser described [here](https://github.com/Princeton-CDH/lenape-timetree/pull/120#issuecomment-1494663926):

> documenting dev steps for the revised approach we agreed on in our whiteboarding session last week:
> 
> - `deselectAll` method should be removed 
> - new `updateSelection` method that sets highlighting in the tree based on the url (one/both/neither of tag param & leaf hash); this method should be called when the page is loaded 
> - select by tag / leaf - rename to `setCurrentLeaf` / `setCurrentTag` (or similar) ; logic is to update url (param or hash) and then call `updateSelection`
> - `closeLeafDetails` - unset current leaf and tag, then call `updateSelection`; I think the proposed approach we discussed was to unset by calling `setCurrentLeaf` / `setCurrentTag` with a null value or similar
> - `closeTag` - unset current tag and then call `updateSelection` 
> - add unit tests for the new methods including interactions between these parameters (both, none, either one)
> 
> Additionally: let's create a new branch `feature/tag-interaction` (or similar) and then merge this and the other tag-related PR into that branch. Dev review can happen before merging into that branch, acceptance testing can happen from a new PR to merge all the tag interactions into `develop`.

related issues:
- #115 
- #114 
- #112 (display active tag in a button above leaf detail panel)
- #111 (when tag is active, highlight leaves with that tag; make leaves without that tag inactive; does not currently handle inactive for tab index / tab navigation)
- #113 (tag stays active until tag or leaf is closed)
- #118 

From Rebecca: @gissoo : please review tag styles and basic interaction; then Jeff can do acceptance testing on all the tag functionality together once we are happy with it.

## revisions after review
- [x] fix x alignment on mobile
- [x] fix spacing between tags ("spacing between the tag containers inside the panel is 12px in the designs but it looks a lot more in the implementation")
- [x] try setting low opacity on container instead of individual leaves
- [x] restore dimmed labels
- [x] handle missing tag name

## acceptance testing 
@jhimpele please test the following on https://lenape-timetree-dev-pr-141.onrender.com/ 

- [x] browse the [tag index](https://lenape-timetree-dev-pr-141.onrender.com/tags/) and select a tag; it should take you to the timetree page with that tag activated
- [x] when you select a tag on the timetree view (either coming from the tag list or by selecting a tag from the leaf details panel), you should see the active tag displayed above the leaf details panel; it should display the readable version and not the slug (i.e., capitalization if any; spaces instead of dashes)
- [x] when a tag is selected, all leaves with that tag should be highlighted; leaves without that tag should be deactivated (dimmer / lower opacity) and not clickable
- [x] closing the active tag (by clicking the X on the active tag indicator) should deselect the tag and return the tree to the normal view; if leaf detail panel is open it should stay open
- [x] closing the leaf detail view when a tag is active should also deselect the tag

### CODE REVIEW

File: assets/js/leaves.js [hunk 1]
```diff
@@ -42,28 +42,42 @@ class Leaf {
...
+      history.replaceState(null, "", urlNoHash);
+    } else {
+      // get the actual leaf target from DOM
+      let target = Leaf.getLeafTarget(event.target);
+      // get leaf ID
+      let leafID = target.dataset.id;
+      // update URL to reflect the currently selected leaf;
+      // replace the location & state to avoid polluting browser history
+      window.location.replace(`#${leafID}`);
+      history.replaceState(null, "", `#${leafID}`);
```
Comment (rlskoeser) [inline assets/js/leaves.js:59]: I think we don't actually need both of these; let's drop the `location.replace` since it doesn't work for changing tags without a reload, and then we can always update the url the same way everywhere, no matter what we're changing.

File: assets/js/leaves.js [hunk 2]
```diff
@@ -42,28 +42,42 @@ class Leaf {
...
+    if (tag == undefined) {
+      history.replaceState(
+        null,
+        "",
+        window.location.pathname + window.location.hash
+      );
+    } else {
+      // set URL to URL of self, with tag updated
+      let url = new URL(window.location.href);
+      url.searchParams.set("tag", tag);
```
Comment (rlskoeser) [inline assets/js/leaves.js:76]: Here also, I think we can consolidate and use consistent syntax - use the `URL` object either way, and just `set` or `delete` the tag value depending on whether tag is set or not.

File: assets/js/leaves.js [hunk 3]
```diff
@@ -130,24 +171,17 @@ class Leaf {
...
-    Leaf.deselectAll();
     panel.parentElement.classList.remove("show-details");
     panel.parentElement.classList.add("closed");
+    Leaf.setCurrentLeaf(undefined);
+    Leaf.closeTag();
+  }
+
+  static closeTag() {
+    // unset current tag and then call updateSelection
+    Leaf.setCurrentTag(undefined);
```
Comment (rlskoeser) [inline assets/js/leaves.js:184]: Looks kind of weird to call these functions with `undefined`, and shouldn't be necessary. I think we get `undefined` if we don't pass a parameter at all - see [mdn on js default parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters). Or we could set a default value if that looks cleaner / more readable to you (or if it's easier to test).

```suggestion
    Leaf.setCurrentLeaf();
    Leaf.closeTag();
  }

  static closeTag() {
    // unset current tag and then call updateSelection
    Leaf.setCurrentTag();
```

**Review by rlskoeser — COMMENTED** [review]
Looks good and is working really well! Have a few suggestions that I hope will be easy to incorporate and will clean up & simplify the code a little bit.

**Comment by gissoo** [comment]
@rlskoeser @quadrismegistus thank you!! Tested on firefox, chrome, and safari desktop and mobile emulator. 

- when selecting a tag the behavior on the tree and the tag container above the panel matches the designs 
- when closing a tag or the panel the behavior matches the designs 
- when selecting related leaves the behavior matches the designs (although I did see some weird behavior that I can't seem to replicate again, where all the leaves with the same tag had outlines and looked "active")

A few problems (style issues)
- is there any way to dim the leaves / disable them visually without bringing down their opacity individually? This looks kind of weird – in Figma I just made the entire group of inactive leaves dim (40% opacity), this way they don't individually lose their opacity. 
Screenshot 2023-04-10 at 3 42 10 PM 

- @rlskoeser you had asked me to align the two "x" close buttons on the mobile designs, but they are not aligned in the implementation. Let me know if you need further specs from Figma on this. 
Screenshot 2023-04-10 at 3 44 59 PM

- I really need to fix the glow soon! The labels look terribly hard to read on them right now. 
Screenshot 2023-04-10 at 3 45 54 PM

**Comment by rlskoeser** [comment]
@gissoo thank you for the careful testing!

Ah, thanks for reminding me about the close buttons not lining up on mobile. I noticed it but then I forgot and lost track.

I'm not sure what you mean about dimming the group of leaves without changing individual leaf opacity. Is the problem where they overlap, or something else? (I did notice the overlap looking weird). Do you mean I can apply the opacity on the `g` container instead of individual leaves? I can try that, but not sure if I can then override that opacity for the leaves that do have the active tag. If opacity doesn't work, then maybe you will need to determine an equivalent color I can apply that doesn't use opacity (but I can try other options first).

Agreed about the glow + readability; wasn't sure what to do about them (it sounded like you have something in mind for improving this).

**Comment by gissoo** [comment]
@rlskoeser I forgot to add, the spacing between the tags inside the container is 12px in the designs but it looks a lot more in the implementation

**Comment by jhimpele** [comment]
regarding 2, do we need to have the selected tag appear above the panel? Seems out of place, and redundant. My opinion is that it would be better to just darken the shade of the selected panel. Also, did we discuss re-selecting the tag in the panel to de-select it, instead of needing the extra panel to do that. 

At least one of the tags labels atop the panel are blank, it showed only an X. And I forgot which one!

Did we decide to not use the dimmed leaf titles while tags are activated? Or is the dimming not yet implemented?

**Comment by rlskoeser** [comment]
@gissoo I've updated based on your comments: fixed the close button alignment, adjusted the spacing between tags (would you believe it was a single space taking up a little bit of extra room?!), added back the labels at lower opacity.  Do you want to check any of this?  

I tried setting the opacity on the container, but it isn't possible for child element opacity to override the opacity of their parents. Moving nodes between layers in the svg just to set opacity is too complicated, I'd rather not attempt. I think we should just leave it as is, but maybe we can try to adjust the layout so leaves don't overlap too much.
